src/function/window/window_custom_aggregator.cpp

```diff
@@ -87,16 +86,17 @@ unique_ptr<WindowAggregatorState> WindowCustomAggregator::GetGlobalState(ClientC
 	return make_uniq<WindowCustomAggregatorGlobalState>(context, *this, group_count);
 }
 
-void WindowCustomAggregator::Finalize(WindowAggregatorState &gstate, WindowAggregatorState &lstate,
-                                      CollectionPtr collection, const FrameStats &stats) {
+void WindowCustomAggregator::Finalize(ExecutionContext &context, WindowAggregatorState &gstate,
+                                      WindowAggregatorState &lstate, CollectionPtr collection,
+                                      const FrameStats &stats) {
 	//	Single threaded Finalize for now
 	auto &gcsink = gstate.Cast<WindowCustomAggregatorGlobalState>();
 	lock_guard<mutex> gestate_guard(gcsink.lock);
 	if (gcsink.finalized) {
 		return;
 	}
 
-	WindowAggregator::Finalize(gstate, lstate, collection, stats);
+	WindowAggregator::Finalize(context, gstate, lstate, collection, stats);
```

Comment:
I wonder if instead of us passing it into all these functions, the `context` could live in the global state or the aggregator? Ofc only where it makes sense - I just chose an arbitrary line for the comment.

Reply:
I did that originally and it caused problems (memory use after free) because the contexts I get at construct time may be from other threads that have moved on to other functions. The window function global and local states aren't tied to pipelines the way operator states are.